lib/puppet/resource/type.rb

```diff
@@ -413,6 +413,9 @@ def assign_defaults(resource, scope, resource_hash)
 
   def validate_resource_hash(resource, resource_hash)
     Puppet::Pops::Types::TypeMismatchDescriber.validate_parameters(resource.to_s, parameter_struct, resource_hash)
+  rescue Puppet::ParseError => e
```

Comment:
I guess I do not understand the state a resource may be in when the problem occurs. If I understand the proposed fix, it catches any problem that is found by parameter validation if the resource has a "consumes" in it during an "environment compilation"). Doesn't that include bona fide errors? 
The fix is also brittle since it is sensitive to the name/impl of the environment compiler.

I would much rather see resources that have the special trait be marked as having a special state ("partially specified") and that such resources only may exist under some conditions and that it is an error to have resources in that state under other conditions (e.g. when finishing a node catalog)